// Copyright 2012 Google Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Author: jmarantz@google.com (Joshua Marantz)
// Protobufs to help with property caching.

syntax = "proto2";

option optimize_for = LITE_RUNTIME;

package net_instaweb;

message PropertyValueProtobuf {
  required string name = 1;
  required bytes body = 2;

  // timestamp of last write -- ms since 1970.
  optional int64 write_timestamp_ms = 3;

  // Bitmask indicating which of the last 64 writes changed the body.
  // Bits are shifted in from LSB to MSB, so the min(num_writes, 64)
  // least significant bits are relevant for computing the ratio of
  // updates.
  optional uint64 update_mask = 4 [ default = 0 ];

  // Total number of writes ever done on this property value.
  optional int64 num_writes = 5 [ default = 0 ];
};

message PropertyCacheValues {
  repeated PropertyValueProtobuf value = 1;
};
